### PR TITLE
Moved UI widgets to a `bevy_widget` crate

### LINKED ISSUE #7190: Move `ButtonBundle`, `ImageBundle`, `TextBundle` and their behaviour into a `WidgetPlugin`
## What problem does this solve or what need does it fill?

During work attempting to implement new and more advanced widgets #7116 & #6517 has uncovered the need to move widget-related functionality into its own plugin for better scalability down the line. 

This is currently one of the issues that prevents faster movement on widget-development of UI.

## What solution would you like?

- create a `WidgetPlugin` that will contain all higher widget-functionality
- Either add this as a sub-plugin under `UiPlugin`
- or create a new crate `bevy_widgets` and have `WidgetPlugin` live there instead and be added to `DefaultPlugins`
    - I'm not sure which would be the better approach.
- Setup a new `Widget` category of examples and move relevant examples there
- Widgets should probably be gated under a feature flag that can be disabled

My hope is that this would lead to a clear separation between the low level building blocks and the specific high-level widget implementation that we end up with. This should also allow users to roll their own widget-implementation while still being able to only rely on the core UI.

## What alternative(s) have you considered?

Keep everything as it is today.

## Additional context

See discussions in the following PRs
- #7116 
- #6517 

## Unanswered Questions

- Should each widget type has its own Plugin as well that would be sub-plugins to `WidgetPlugin`? 
  - Example: `TextWidgetPlugin`, `ButtonWidgetPlugin`, `SliderWidgetPlugin`

### PR DESCRIPTION
**Pull request by Pietrek14**
# Objective

Fixes #7190.

## Solution

- Added a new `bevy_widget` crate to Bevy
- Moved the widgets to the new crate
- Created a seperate category for widget-related examples

---

## Changelog

- Added
  - A `bevy_widget` crate
  - `WidgetPlugin`
- Moved
  - Widgets to `bevy_widget` crate
  - Widget-related examples to a dedicated directory

## Migration Guide
- Widget imports now have to reference the `widget` module instead of `ui`
- If you don't use the default features and want to have widgets available, you will have to enable the `bevy_widget` feature flag
- If you don't use the default plugins and want to have widgets available, you will have to add a `WidgetPlugin` to your app

### CODE REVIEW

File: crates/bevy_widget/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+mod widget;
```
Comment (alice-i-cecile) [inline crates/bevy_widget/src/lib.rs:1]: Can you add forbid(unsafe_code) and warn(missing_docs) crate level rules please?

Once that's in, we should add basic doc strings.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

**Comment by alice-i-cecile** [comment]
Marked as Controversial since this adds a new crate, but it seems like there's pretty clear consensus on splitting apart the code like this.

File: crates/bevy_widget/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#![forbid(unsafe_code)]
+#![warn(missing_docs)]
+//! This crate contains UI widgets like buttons
```
Comment (alice-i-cecile) [inline crates/bevy_widget/src/lib.rs:3]:
```suggestion
//! The official collection of user interface widgets for `bevy_ui`.
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Feedback on structure:

1. I think that the widget bundles belong in the same file as the logic for that widget.
2. I don't think that the `widget` module is useful since we're in a dedicated crate for widgets. We can just make each widget a top level module.

File: crates/bevy_widget/src/lib.rs (hunk 1 above)
Comment (Pietrek14) [reply crates/bevy_widget/src/lib.rs]: The only thing with missing docs is `QueuedText`. It's only used as a `Local` for `text_system`, so should we even keep it public?

Comment (alice-i-cecile) [reply crates/bevy_widget/src/lib.rs]: We should probably keep text_system public, so we need to keep the Local type public or Rust will complain.

**Review by alice-i-cecile — APPROVED** [review]
LGTM now :) I trust that you'll get basic docs for `QueuedText` working.

**Review by Weibye — APPROVED** [review]
💯

**Review by nicopap — APPROVED** [review]
LGTM.

One thing to note is that it cements the distinction between a "render ui" layer and a more abstract "widget" layer for ui. In this split, widgets are bundles (in the future, they are probably going to be more than a collection of components, and be more like a `Prefab` of sort, encapsulating a hierarchy of entities) while the ui primitives are components defined in `bevy_ui`.

I don't see any issue with this. I think UI widgets are necessarily more complex than render components, and the distinction will keep making sense in the future. It also encourages splitting the widget impls from the primitive impls, maybe resulting in more re-usable primitives.

**Comment by alice-i-cecile** [comment]
> It also encourages splitting the widget impls from the primitive impls, maybe resulting in more re-usable primitives.

Agreed, this is my hope too.

**Comment by Pietrek14** [comment]
How do I fix the `check-bans` failing?

**Comment by alice-i-cecile** [comment]
That's fixed via https://github.com/bevyengine/bevy/blob/main/deny.toml.

I don't think it's caused by your PR though: looks like we got a minor version update from `gpu-allocator`.

If you want to fix it, make a new PR with an updated allow rule to ignore this warning.

**Comment by james7132** [comment]
Might be a good idea to use `bevy_widgets` instead, which is one of the currently reserved crates, while `bevy_widget` isn't.

**Comment by mockersf** [comment]
`bevy_ui_widgets`, or do we plan to also add 3d widgets to the same crate?

**Comment by Pietrek14** [comment]
From what I can see, both `bevy_widgets` and `bevy_widget` are reserved.

Also, if we find it necessary to seperate 3d widgets into another crate, we can call it `bevy_3d_widget` or sth like this and keep the "regular" widgets in the `bevy_widget` crate.

**Comment by alice-i-cecile** [comment]
Yeah, I would call the 3D ones "gizmos".

**Comment by mockersf** [comment]
Disclaimer: I'm not Cart

I think it's too early to do this, and I haven't seen either in this PR or in the issue it fixes a good enough justification for the change. This is just a code organisation changes that doesn't improve anything or unlock anything, as such it's very low priority

**Comment by richchurcher** [comment]
Backlog cleanup: closing due to inactivity. We've already split out a number of crates, so I guess there's nothing to stop us doing this later but probably easier to begin again after 0.15.

**Comment by alice-i-cecile** [comment]
I still think we should do something like this, but yeah, this requires consensus and broad maintainer buy-in :(

**Comment by richchurcher** [comment]
I hovered over the button, but figured someone would yell if they really wanted it kept!

**Pull request closed without merging**